lib/doorkeeper/models/access_token_mixin.rb

```diff
@@ -191,10 +191,23 @@ def find_or_create_for(application, resource_owner_id, scopes, expires_in, use_r
       # @return [Doorkeeper::AccessToken] array of matching AccessToken objects
       #
       def authorized_tokens_for(application_id, resource_owner_id)
-        ordered_by(:created_at, :desc)
+        # In case of static token expiry times, we can optimize fetching tokens
+        # by already filtering out tokens whose TTL has expired.
+        # We can't do that if custom expirations times have been configured, though.
+        unless Doorkeeper.configuration.option_defined?(:custom_access_token_expires_in)
+          valid_from = Doorkeeper.configuration.access_token_expires_in.seconds.ago
+        end
+
+        tokens = ordered_by(:created_at, :desc)
```

Comment:
As mentioned in https://github.com/doorkeeper-gem/doorkeeper/issues/1314#issuecomment-540625227, I think using time math would handle every scenario.

We have `expires_in` stored in the database, so something like this could work:

```
tokens
  .where(
    application_id: application_id,
    resource_owner_id: resource_owner_id,
    revoked_at: nil,
  )
  .where("created_at + expires_in * interval '1 second' > ?", Time.now.utc)
```

I'm not sure if this syntax covers every supported database.

Reply:
> I'm not sure if this syntax covers every supported database.

We must support all the dbs that are supported by ActiveRecord. So either implement such query our self in the adapters , either .. I don't know..

Reply:
Double checked. This syntax won't work in sqlite3. Unsure how to do this. Creating the `expires_at` column would sidestep this whole issue, but I see why we'd want to avoid the redundancy. I'm guessing we're storing `expires_in` because that's the recommended way of reporting the expiration in oauth (https://tools.ietf.org/html/rfc6749#section-4.2.2) but maybe `expires_in` should be calculated from `expires_at`, instead of the other way round? Maybe this has been discussed before.

Reply:
Something like this may work (have not tested):

```ruby
tokens
  .where(
    application_id: application_id,
    resource_owner_id: resource_owner_id,
    revoked_at: nil,
  )
  .where(
    if ActiveRecord::Base.connection.adapter_name == "SQLite"
      "datetime(created_at, '+' || expires_in || ' seconds') > ?"
    else
      "created_at + expires_in * interval '1 second' > ?"
    end,
    Time.now.utc,
  )
```
Thoughts?

Reply:
@EugeneFractal

It would work, yep. But I'm afraid of timezones used by databases (Postgres, Mysql, etc). Any project could set their own local timezones both for project and DBMS so we could have invalid queries. We can try to convert expressions like `created_at + expires_in * interval '1 second'` to UTC in the database, but every database has it's own syntax for this.

Reply:
Something generic but very dangerous would be:

```ruby
def expired_for(application_id, resource_owner_id)
  where.not(expires_at: nil)
    .where(
      application_id: application_id,
      resource_owner_id: resource_owner_id,
      revoked_at: nil,
    )
    .where("#{expiration_date_sql_expression} <= ?", Time.now.utc)
end

def expiration_date_sql_expression
  case ActiveRecord::Base.connection.adapter_name.downcase
  when "sqlite", "sqlite3"
    Arel.sql("DATETIME(created_at, '+' || expires_in || ' SECONDS')")
  when "postgis", "postgresql"
    Arel.sql("created_at + expires_in * INTERVAL '1 SECOND' AT TIME ZONE 'UTC'")
  when "mysql", "mysql2"
    Arel.sql("created_at + expires_in * INTERVAL '1 SECOND'")
  when "sqlserver"
    Arel.sql("DATEADD(second, expires_in, created_at) AT TIME ZONE 'UTC'")
  when "oracleenhanced"
    Arel.sql("created_at + INTERVAL to_char(expires_in) second")
  else
    raise "Unsupported database!"
  end
end
```